### ISSUE TITLE
Md5 hash does not match for two different runs

### ISSUE THREAD

**Issue by Eic41** [doxygen/doxygen#7706]
Hello, since the switch to version 1.8.18, I get "md5 hash does not match for two different runs of C:/Users/amoyer/AppData/Local/Temp/OptimRem/html/class_optim_rem_1_1_calcul_1_1_i_1_1_calculs_dirigeant_a07709f8061d8aed6c520686ca3d13d81_cgraph.dot!" errors.
I have attached a project to you that I cleaned up with the .doxyfile file that I use. You should be able to reproduce the bug.

In the meantime, I abandoned version 1.8.18 to reuse version 1.8.15.

[Project.zip](https://github.com/doxygen/doxygen/files/4491411/Project.zip)

**Comment by albert-github** [comment]
Looks like that the (merged) commit:
> Commit: 572035ffcd0e4f7313bab23d375d63008ebbe5a3 [572035f]
> Parents: 047df30104
> Date: Tuesday, May 14, 2019 9:45:27 AM
> Commit Date: Monday, June 10, 2019 1:46:48 PM
> delayed creation and update of .md5 files after successful creation of output files

still works but that the next working (merged) commit:
> Commit: 7bae1f4e7c1f7b38e205f158cc5cbe0e4b956b75 [7bae1f4]
> Parents: 182a5e8af0
> Date: Friday, May 24, 2019 8:33:33 AM
> Commit Date: Thursday, June 13, 2019 8:55:20 PM
> refactoring dot.cpp

doesn't work anymore.


@doxygen, @thomas-haller any idea's

**Comment by thomas-haller** [comment]
Currently I do not have much time to investigate in this. What I have found so far:
Last year the problem was that the node numbering in the dot files changed and caused this problem. This time the ordering of the graph content changed.
First run the order is
GarantieChomage,AssuranceVie,PrevoyanceMadelin,ChomageMadelin,Perp,Art82,ComplSante
Second run the order is
GarantieChomage,AssuranceVie,ChomageMadelin,Perp,PrevoyanceMadelin,Art82,ComplSante
What can change this ordering?

**Comment by albert-github** [comment]
@thomas-haller thanks for this analysis.

You are referring to the method ` CalculerCocheOptions` in the file  uCalculsDirigeant.cpp and uCalculsEntreprise.cpp. In the first file the calls are present in the second file the method is empty but here we have also a: `using CalculsDirigeant::CalculerCocheOptions;`.

I did a search for all usages etc. of CalculerCocheOptions and found:
```
OptimRem.Calcul.I/uCalculsDirigeant.cpp:                        void __fastcall CalculsDirigeant::CalculerCocheOptions(Dirigeant * pDirigeant)
OptimRem.Calcul.I/uCalculsDirigeant.h:                          void __fastcall CalculerCocheOptions(Dirigeant* pDirigeant);
OptimRem.Calcul.I/uCalculsEntreprise.cpp:                       void __fastcall CalculsEntreprise::CalculerCocheOptions(Entreprise *pEntreprise)
OptimRem.Calcul.I/uCalculsEntreprise.h:                         using CalculsDirigeant::CalculerCocheOptions; //Afin que cette méthode (parente) puisse encore être appelée.
OptimRem.Calcul.I/uCalculsEntreprise.h:                         void __fastcall CalculerCocheOptions(Entreprise *pEntreprise); //Override de celle parente
```

No idea why order is different and why the function is listed in uCalculsEntreprise.cpp unless this is an effect of the `using` part (I'm not an expert in `using`).

**Comment by doxygen** [comment]
I've made a fix. Please try it and let me know if it solves the problem.

Background: when sorting, the name (without scope) was taken into account and if they were the same the line number of the definition was used to sort. But these numbers could also be the same if the definitions come from different files. So now I also added the filename to the sorting criteria as well (and used the same sorting function for lists and sdicts).

**Comment by albert-github** [comment]
Unfortunately it doesn't solve all the problems.

Based on this issue I searched the Fossies warnings files and found a couple of packages that have the same problem.
I tried the https://github.com/bareos/bareos/archive/Release/19.2.7.tar.gz package with the following doxygen configuration file: [Doxyfile.tar.gz](https://github.com/doxygen/doxygen/files/4496378/Doxyfile.tar.gz)
Doxygen version: 1.8.19 (f3ebc866c88b31ab393637e30189d41ddb7f84f0*)

The number of messages went from 48 to 40 (I killed the current process as it wants to convert 20118 dot images and as far as I can see all messages are already in the warning file).
The total process for this build is quite fast.

**Comment by doxygen** [comment]
Some background info:

In the example provided by Albert, there were static functions with the same signature defined in different files (belonging to different plugins). The odd thing is that all functions were defined in the same namespace, which does not make much sense for static functions. 

This construct caused the callgraphs not to have unique names (and thus overwriting each other). I did not really fix this, but removed the part where the dot file was already removed before the end of a doxygen run. So now every run you'll still see some dot images being regenerated, fixing this would break compatibility with existing code (links will not longer be the same).

Another issue that I found and fixed, is that in some cases for each run the SVG file was patched and an additional "target" attribute was added for a link, causing an invalid SVG syntax. I now check if the target attribute is already there before adding it.

**Comment by albert-github** [comment]
I don't know the bareos package, but the static functions might be from backends (etc.) and probably only one is allowed, when doing a brute force doxygen approach on it (i.e. taking all sources) everything is taken and thus resulting in the mentioned dproblems.

I'll have a go at it, it will take a while to test the packages I found ....

**Comment by albert-github** [comment]
I reran the bareos 19.2.7 package, but I still get 40 times the message.
Doxygen version: 1.8.19 (cd581388f3d013c501e3cefbaf3e81cf93d46fcb)
Before running I removed the old html directory.
```
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a6d006ea832ebf5de6fb33296ca6303fc_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a6d006ea832ebf5de6fb33296ca6303fc_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a6d006ea832ebf5de6fb33296ca6303fc_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a6d006ea832ebf5de6fb33296ca6303fc_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_afd5b669e8c378785035dd2d03776bf69_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_afd5b669e8c378785035dd2d03776bf69_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a6d40b3689da483075f6f8e82ab3a18c4_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a6d40b3689da483075f6f8e82ab3a18c4_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a4d077f18e8af9aee60560a9bb64e09df_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_ae9621cf108d9f267f0bf5d419323063c_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_ad0fd5e0968929ef90cc8a723f8c3d98e_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_ad0fd5e0968929ef90cc8a723f8c3d98e_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a8cd5a90716ce6f4ac9325f6d4506582f_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a8cd5a90716ce6f4ac9325f6d4506582f_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a8cd5a90716ce6f4ac9325f6d4506582f_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a8cd5a90716ce6f4ac9325f6d4506582f_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a8cd5a90716ce6f4ac9325f6d4506582f_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_abe972255041866d16324e27e69a91223_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_abe972255041866d16324e27e69a91223_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_abe972255041866d16324e27e69a91223_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_abe972255041866d16324e27e69a91223_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_abe972255041866d16324e27e69a91223_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_acdb9aa5375a712d930ceb8df54c6f05e_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_acdb9aa5375a712d930ceb8df54c6f05e_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_acdb9aa5375a712d930ceb8df54c6f05e_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_acdb9aa5375a712d930ceb8df54c6f05e_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_adb77b7d69d680cf8800db2f31092d2c3_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_adb77b7d69d680cf8800db2f31092d2c3_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a0a8433edb3417c5aae46058c7bc3e87c_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a6ab0ee850fc12673faf2c324429adceb_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_af5e6a04250d89bbd5ccdcb4ff128d328_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_af5e6a04250d89bbd5ccdcb4ff128d328_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a8ff00bbe18c4d857e734611825c05c45_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a8ff00bbe18c4d857e734611825c05c45_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_a1de6eec693022d8c602635f4c7ab1ecc_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_ac90e6f7c946bcd14164f79f711f74118_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_ac90e6f7c946bcd14164f79f711f74118_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_ac90e6f7c946bcd14164f79f711f74118_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_ac90e6f7c946bcd14164f79f711f74118_cgraph.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/bareos-Release-19.2.7/html/namespacefiledaemon_ab503cf3178bc58c52cdf5fae4289a4fa_cgraph.dot !
```

I also found for the https://sourceforge.net/projects/upp/files/upp/2019.2/upp-x11-src-13664.tar.gz package (The used configuration file: [Doxyfile.tar.gz](https://github.com/doxygen/doxygen/files/4499104/Doxyfile.tar.gz)) a number of message:
```
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/upp-x11-src-13664/html/_2_dispatcher_test_8h__incl.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/upp-x11-src-13664/html/_2_dispatcher_test_8h__dep__incl.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/upp-x11-src-13664/html/_2_dispatcher_test_8h__incl.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/upp-x11-src-13664/html/_2_dispatcher_test_8h__dep__incl.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/upp-x11-src-13664/html/_2_dispatcher_test_8h__incl.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/upp-x11-src-13664/html/_2_dispatcher_test_8h__dep__incl.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/upp-x11-src-13664/html/_2_dispatcher_test_8h__incl.dot !
error: md5 hash does not match for two different runs of /cygdrive/e/Fossies/check_md5/upp-x11-src-13664/html/_2_dispatcher_test_8h__dep__incl.dot !
```

**Comment by albert-github** [comment]
Seen the problem I see with the upp package it looks like that it is only for some include graphs. Would it necessary to have some sorting here as well?

**Comment by doxygen** [comment]
@albert-github I haven't been able to reproduce the problem with package upp on my end.

**Comment by albert-github** [comment]
@doxygen 
- Strange I see the problem with Cygwin, I'll try to create a little bit a smaller example (I already succeeded to go from 21K dot files to 5351 dot files.
- On windows I don't get that far with te original input as I get the message:
   > Generating code for file bazaar/DXF/Block.cpp...
   > out of dynamic memory in defargsYYlex()
   > Exiting...

   this is a different problem and needs to be investigated further ...

**Comment by albert-github** [comment]
I found the basic problem regarding the upp package.
When only having as `INPUT`:
```
INPUT                  =  \
./bazaar/Dispatcher \
./bazaar/DispatcherTest \
./bazaar/DispatcherTest2 \
./bazaar/DispatcherTest3 \
./bazaar/DispatcherTest4 \
./bazaar/DispatcherTest5 \
./bazaar/DispatcherTest6
```
we get the messages and we see in the directories:
```
bazaar/DispatcherTest:
DispatcherTest.h  DispatcherTest.lay  DispatcherTest.upp  main.cpp

bazaar/DispatcherTest2:
DispatcherTest.h  DispatcherTest.lay  DispatcherTest2.upp  main.cpp

bazaar/DispatcherTest3:
DispatcherTest.h  DispatcherTest.lay  DispatcherTest3.upp  main.cpp

bazaar/DispatcherTest4:
DispatcherTest.h  DispatcherTest.lay  DispatcherTest4.upp  main.cpp

bazaar/DispatcherTest5:
DispatcherTest.h  DispatcherTest.lay  DispatcherTest5.upp  main.cpp

bazaar/DispatcherTest6:
DispatcherTest.h  DispatcherTest.lay  DispatcherTest6.upp  main.cpp
```
so similar code and similar file names  / incluides resulting in the problems with "error: md5 hash does not match for two different runs". I also can reproduce the problem under windows. @doxygen it is a bit strange that this problem does not show up at your side.

**Comment by doxygen** [comment]
@albert-github I didn't look into the Doxyfile very carefully and was expecting the warnings to appear on `stdout` instead of in `warnings.log`. Now I see them. One mystery solved.

**Comment by doxygen** [comment]
Found the problem; in case two files with the same name where located in different directories where the name of one directory was a prefix of the other, the routine that computed the common prefix was wrong, because it took the last path separator into account.

Example case:
```
/some/path/long/file.h
/some/path/longer/file.h
```
Then the output files could overwrite each other.

Let me know if you still see issues after pulling in this commit.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
